framework/src/binary_manager/binary_manager_get_state.c

```diff
@@ -0,0 +1,65 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/***************************************************************************
+ * Included Files
+ ***************************************************************************/
+
+#include <debug.h>
+#include <string.h>
+#include <tinyara/binary_manager.h>
+#include <binary_manager/binary_manager.h>
+#include "binary_manager_internal.h"
+
+int binary_manager_get_state(char *binary_name)
+{
+	int ret;
+	binmgr_request_t request_msg;
+	binmgr_getstate_response_t response_msg;
+
+	if (binary_name == NULL || strlen(binary_name) > BIN_NAME_MAX - 1) {
+		bmdbg("get_binary_info failed : invalid param.\n");
+		return BINMGR_INVALID_PARAM;
+	}
+
+	ret = binary_manager_set_request(&request_msg, BINMGR_GET_STATE, binary_name);
+	if (ret < 0) {
+		return BINMGR_COMMUNICATION_FAIL;
+	}
+
+	ret = binary_manager_send_request(&request_msg);
+	if (ret < 0) {
+		bmdbg("Failed to send request msg %d\n", ret);
+		return BINMGR_COMMUNICATION_FAIL;
+	}
+
+	ret = binary_manager_receive_response(&response_msg, sizeof(binmgr_getstate_response_t));
+	if (ret < 0) {
+		bmdbg("Failed to receive response msg %d\n", ret);
+		return BINMGR_COMMUNICATION_FAIL;
+	}
+
+	if (response_msg.result == BINMGR_OK) {
+		ret = response_msg.state;
+	} else {
+		bmdbg("Binary manager getstate FAIL %d\n", response_msg.result);
+		ret = response_msg.result;
```

Comment:
If defined values for state(guess binary_state_e) include error case, you don't have to use structure, enum will be enough.
Please make sure that name of this function is 'get_state'
So, here you got message through the mq from lower, you should parse it and return newly defined values in this framework to requested layer.
To sum up,
1. Define state values in framework.
2. Here we need to parse message from lower layer, and return newly defined values to user.
This will help upper layer to reduce reference code, that's why we make framework, actually.

At least, here you should return error value of framework defined, not result itself.
Currently this returns values in framework or msg from lower layer...it's not good.

Reply:
Yes. I agree with your opinion. Currently all values are defined in <tinyara/binary_manager.h> and user and kernel use them. so user should include that file for binary manager usage.
But what I worried about is that there are duplicated definitions in framework and kernel if they are defined in framework also.

Reply:
You'd better divide values. do not put all values to <tinyara/binary_manager.h>.
Framework is necessary to provide small set of reference code to above layer to make their application more easy & efficiently.
So put some values for user to some headers in framework/include, application doesn't have to include <tinyara/binary_manager.h> because they do not need all values in there.
Below is sample of re-definition of lower values. (lower has 12 of PCM_FORMAT but app doesn't need it)
```
typedef enum audio_format_type_e {
	/* Signed 8 bit */
	AUDIO_FORMAT_TYPE_S8 = PCM_FORMAT_S8,
	/* Signed 16 bit use this as a default */
	AUDIO_FORMAT_TYPE_S16_LE = PCM_FORMAT_S16_LE,
	/* Signed 32 bit */
	AUDIO_FORMAT_TYPE_S32_LE = PCM_FORMAT_S32_LE
} audio_format_type_t;
```